cirq/ops/thermal_relaxation_channel.py

```diff
@@ -0,0 +1,219 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Module for the thermal relaxation channel."""
+
+from typing import Iterable
+
+import numpy as np
+
+from cirq import protocols, value
+from cirq.ops import gate_features
+
+
+@value.value_equality
+class ThermalRelaxationChannel(gate_features.SingleQubitGate):
+    """Dampen qubits at varying rates through non ideal disipation.
+
+    This channel models the effect of energy dissipation into the environment
+    as well as the environment depositing energy into the system at varying
+    rates.
+    """
+
+    def __init__(self, p_exchange: float, p_relaxation: float,
+                 p_dephasing: float) -> None:
+        r"""The thermal relaxation channel.
+
+        Construct a channel to model energy dissipation into the environment
+        as well as the environment depositing energy into the system. The
+        probabilities with which the energy exchange occur are given by
+        `p_relaxation` and `p_dephasing` with the probability of the environment
+        being not excited is given by `p_exchange`.
+
+        Like the `GeneralizedAmplitudeDampingChannel`, the stationary state of
+        this channel is the diagonal density matrix with probability
+        `p_exchange` of being |0⟩ and probability `1-p_exchange` of being |1⟩.
+
+        We begin by examining the `GeneralizedAmplitudeDampingChannel`'s choi
+        matrix:
```

Comment:
This should be the Choi matrix of the thermal relaxation channel. The relationship to GAD shouldn't be the focus of the docstring. We should point it out *after* we define the thermal relaxation channel.

Reply:
Fixed the matrix. Also I can't seem to reply to your below question, but there are valid channels produced when `p_exchange != p_relaxation + p_excitation`